### ISSUE TITLE
BigQuery Storage API periodically hangs while reading table data

### ISSUE THREAD

**Issue by almegeddon** [googleapis/java-bigquerystorage#531]
#### Environment details

**OS type and version:**

-bash-4.2$ cat /etc/os-release
NAME="Amazon Linux AMI"
VERSION="2018.03"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2018.03"
PRETTY_NAME="Amazon Linux AMI 2018.03"
ANSI_COLOR="0;33"
CPE_NAME="cpe:/o:amazon:linux:2018.03:ga"
HOME_URL="http://aws.amazon.com/amazon-linux-ami/"

**Java version:**

-bash-4.2$ java -version
openjdk version "11.0.8" 2020-07-14 LTS
OpenJDK Runtime Environment Corretto-11.0.8.10.1 (build 11.0.8+10-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.8.10.1 (build 11.0.8+10-LTS, mixed mode)

**bigquerystorage version(s):**

```
[INFO] +- com.google.cloud:google-cloud-bigquerystorage:jar:1.4.1:compile
[INFO] |  +- io.grpc:grpc-api:jar:1.31.1:compile
[INFO] |  +- org.codehaus.mojo:animal-sniffer-annotations:jar:1.19:runtime
[INFO] |  +- io.grpc:grpc-stub:jar:1.31.1:compile
[INFO] |  +- io.grpc:grpc-protobuf:jar:1.31.1:compile
[INFO] |  +- io.grpc:grpc-protobuf-lite:jar:1.31.0:compile
[INFO] |  +- com.google.auto.value:auto-value-annotations:jar:1.7.2:compile
[INFO] |  +- com.google.api.grpc:proto-google-cloud-bigquerystorage-v1alpha2:jar:0.104.1:compile
[INFO] |  +- com.google.api.grpc:proto-google-cloud-bigquerystorage-v1beta1:jar:0.104.1:compile
[INFO] |  +- com.google.api.grpc:proto-google-cloud-bigquerystorage-v1beta2:jar:0.104.1:compile
[INFO] |  +- com.google.api.grpc:proto-google-cloud-bigquerystorage-v1:jar:1.4.1:compile
[INFO] |  +- com.google.api:gax-grpc:jar:1.58.2:compile
[INFO] |  +- io.grpc:grpc-auth:jar:1.31.1:compile
[INFO] |  +- io.grpc:grpc-netty-shaded:jar:1.31.1:compile
[INFO] |  +- io.grpc:grpc-alts:jar:1.31.1:compile
[INFO] |  +- io.grpc:grpc-grpclb:jar:1.31.1:compile
[INFO] |  +- org.conscrypt:conscrypt-openjdk-uber:jar:2.2.1:compile
[INFO] |  +- org.apache.commons:commons-lang3:jar:3.5:compile
[INFO] |  +- org.json:json:jar:20200518:compile
[INFO] |  +- io.grpc:grpc-core:jar:1.31.1:compile
[INFO] |  +- com.google.android:annotations:jar:4.1.1.4:runtime
[INFO] |  +- io.perfmark:perfmark-api:jar:0.19.0:runtime
[INFO] |  +- com.google.api.grpc:grpc-google-cloud-bigquerystorage-v1beta1:jar:0.104.1:compile
[INFO] |  +- com.google.api.grpc:grpc-google-cloud-bigquerystorage-v1beta2:jar:0.104.1:compile
[INFO] |  \- com.google.api.grpc:grpc-google-cloud-bigquerystorage-v1:jar:1.4.1:compile
```

#### Steps to reproduce

Issue presents intermittently while reading from one of our tables with approximately 150 million rows:

projects/dev-10450-26/datasets/optimize/tables/target_file_full

Approximately 95% of rows are read in under 10 minutes. The remaining readers appear to stall with next to no throughput taking around 90 minutes to read the remaining 5% of rows.

When this issue does not present, we are able to read the entire table in around 10 minutes. Our reader is based on this API's sample reader implementation, modified to kick off a thread to consume all available streams.

#### Code example

```java
import com.google.api.gax.rpc.ServerStream;
import com.google.cloud.bigquery.storage.v1.*;
import com.google.common.base.Preconditions;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorLoader;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ReadChannel;
import org.apache.arrow.vector.ipc.message.MessageSerializer;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.Schema;
import org.apache.arrow.vector.util.ByteArrayReadableSeekableByteChannel;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class BigQueryReader {
    private final ReadBigQueryConfig config;

    public BigQueryReader(ReadBigQueryConfig config) {
        this.config = config;
    }

    public void read() throws Exception {
        try (BigQueryReadClient bqr = BigQueryReadClient.create()) {
            System.out.println("Reading rows via BQ storage API from: " + config.getTable());

            ReadSession.TableReadOptions options = ReadSession.TableReadOptions.newBuilder()
                .addAllSelectedFields(config.getColumns())
                .build();

            ReadSession.Builder sessionBuilder = ReadSession.newBuilder()
                .setTable(config.getTable())
                .setDataFormat(DataFormat.ARROW)
                .setReadOptions(options);

            CreateReadSessionRequest request = CreateReadSessionRequest.newBuilder()
                .setParent(config.getParent())
                .setReadSession(sessionBuilder)
                .setMaxStreamCount(0)
                .build();

            ReadSession session = bqr.createReadSession(request);
            Preconditions.checkState(session.getStreamsCount() > 0);
            System.out.println(session.getStreamsCount() + " streams available for reading");

            AtomicInteger readerCount = new AtomicInteger(1);
            List<CompletableFuture<Void>> futures = session.getStreamsList().stream()
                .map(rs -> CompletableFuture.runAsync(() -> {
                    String readerName = config.getTableId() + "_reader_" + readerCount.getAndIncrement();
                    try (SimpleRowReader reader = new SimpleRowReader(readerName, session.getArrowSchema())) {
                        ReadRowsRequest readRowsRequest = ReadRowsRequest.newBuilder()
                            .setReadStream(rs.getName())
                            .build();

                        // Process each block of rows as they arrive and decode using our simple row reader.
                        ServerStream<ReadRowsResponse> stream = bqr.readRowsCallable().call(readRowsRequest);
                        for (ReadRowsResponse response : stream) {
                            Preconditions.checkState(response.hasArrowRecordBatch());
                            reader.processRows(response.getArrowRecordBatch());
                        }
                    } catch (IOException iox) {
                        throw new IllegalStateException("Problem reading rows: " + iox.getMessage(), iox);
                    }
                }))
                .collect(Collectors.toList());
            CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).join();
        }
    }

    private static class SimpleRowReader implements AutoCloseable {
        private final BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
        private final String name;
        private final VectorSchemaRoot root;
        private final VectorLoader loader;
        private int totalBatches;
        private int totalRows;

        public SimpleRowReader(String name, ArrowSchema arrowSchema) throws IOException {
            this.name = name;
            Schema schema =
                MessageSerializer.deserializeSchema(
                    new ReadChannel(
                        new ByteArrayReadableSeekableByteChannel(
                            arrowSchema.getSerializedSchema().toByteArray()
                        )
                    )
                );
            Preconditions.checkNotNull(schema);

            List<FieldVector> vectors = new ArrayList<>();
            for (Field field : schema.getFields()) {
                vectors.add(field.createVector(allocator));
            }
            root = new VectorSchemaRoot(vectors);
            loader = new VectorLoader(root);
        }

        public void processRows(ArrowRecordBatch batch) throws IOException {
            org.apache.arrow.vector.ipc.message.ArrowRecordBatch deserializedBatch =
                MessageSerializer.deserializeRecordBatch(
                    new ReadChannel(
                        new ByteArrayReadableSeekableByteChannel(
                            batch.getSerializedRecordBatch().toByteArray()
                        )
                    ), allocator
                );

            loader.load(deserializedBatch);
            deserializedBatch.close();
            writeBatch();
            root.clear();
        }

        private void writeBatch() {
            totalBatches++;
            // we convert rows into our own custom objects here and increment row count...
        }

        @Override
        public void close() {
            root.close();
            allocator.close();
            System.out.println(String.format(
                "BQ Reader %s stats - Total Batches: %,d; Total Rows: %,d", name, totalBatches, totalRows));
        }
    }
}
```

**Comment by twinkle-kadia** [comment]
I have written BigQueryReader class. and  tested couple of times but not able reproduce. Please give me your suggestions.

**Comment by almegeddon** [comment]
@twinkle-kadia how large is the table you're reading from? We have one table where we have observed this behavior multiple times (we read from several others w/ out issue, but they are much smaller). This table is ~ 26GB with 150M rows. As mentioned previously, it presents as if one or two of the readers get hung up as the majority of rows are read in really fast.

I noticed this table is consistently allocated 25 stream readers when I let BQ allocate the streams via `CreateReadSessionRequest.setMaxStreamCount(0)` I've tried playing around with this (set it to 20, 15) which didn't seem to make a difference.

**Comment by twinkle-kadia** [comment]
I have tested on large size table. My table size is 35GB with around 313M rows. I have tested multiple times with changing setMaxStreamCount() values, tried with 5,10 & 20 values, but it is executed without hang up or fail. And taken times as below :
Size	        Rows	StreamCountValue	   Time
35GB	313M 	5			            58 Mins
35GB	313M 	10			            50 Mins
35GB	313M 	20			            40 Mins
